pkg/verify/verify.go

```diff
@@ -283,9 +223,54 @@ func (v *releaseVerifier) Verify(ctx context.Context, releaseDigest string) erro
 		}
 		return fmt.Errorf("unable to locate a valid signature for one or more sources")
 	}
+
+	v.cacheVerification(releaseDigest, signedWith)
+
 	return nil
 }
 
+// Signatures returns a copy of any cached signatures that have been validated
+// so far. It may return no signatures.
+func (v *ReleaseVerifier) Signatures() map[string][][]byte {
+	copied := make(map[string][][]byte)
+	v.lock.Lock()
+	defer v.lock.Unlock()
+	for k, v := range v.signatureCache {
+		copied[k] = v
+	}
+	return copied
+}
+
+// hasVerified returns true if the digest has already been verified.
+func (v *ReleaseVerifier) hasVerified(releaseDigest string) bool {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+	_, ok := v.signatureCache[releaseDigest]
+	return ok
+}
+
+const maxSignatureCacheSize = 64
+
+// cacheVerification caches the result of signature check for a digest for later retrieval.
+func (v *ReleaseVerifier) cacheVerification(releaseDigest string, signedWith [][]byte) {
+	v.lock.Lock()
+	defer v.lock.Unlock()
+
+	if len(signedWith) == 0 || len(releaseDigest) == 0 || v.signatureCache == nil {
+		return
+	}
+	// remove the new entry
+	delete(v.signatureCache, releaseDigest)
+	// ensure the cache doesn't grow beyond our cap
+	for k := range v.signatureCache {
+		if len(v.signatureCache) < maxSignatureCacheSize {
+			break
+		}
+		delete(v.signatureCache, k)
```

Comment:
This is pretty basic, and could lead to flapping two recent dogests while we hold a bunch of ancient digests in the cache.  Is there some LRU thing we can pull off a shelf?

Reply:
I guess what we want is time aware least recently used (TLRU) eviction, with the time component being min(sig expiration, key expiration).

Reply:
Map walk is random.  If you get flapping you should go play the lottery and retire on it.

I don't think we need anything more than random.

Reply:
From [a 2013 Go comment][1]:

> Don't rely on the ordering for ranges over maps.  Relying on it to be random is just as
bad as relying on it to be sorted, or repeatable, or whatever.  Making it random is an
attempt to *prevent* people from relying on the order, not an *invitation* to rely on
the order for randomness.

Is that still current?  [The spec][2] has:

> The iteration order over maps is not specified and is not guaranteed to be the same from one iteration to the next.

without committing to a particular quality of randomness.

[1]: https://github.com/golang/go/issues/5362#issuecomment-66078617
[2]: https://golang.org/ref/spec#RangeClause